paymentapi/src/test/java/fi/hel/verkkokauppa/payment/api/PaytrailPaymentControllerUnitTests.java

```diff
@@ -0,0 +1,417 @@
+        ReflectionTestUtils.setField(paymentPaytrailService, "paymentItemRepository", paymentItemRepository);
+        ReflectionTestUtils.setField(paymentPaytrailService, "payerRepository", payerRepository);
+    }
+
+    @Test
+    public void whenCreateFromOrderWithValidDataThenReturnStatus201() throws Exception {
+        GetPaymentRequestDataDto paymentRequestDataDto = new GetPaymentRequestDataDto();
+        paymentRequestDataDto.setMerchantId("01fde0e9-82b2-4846-acc0-94291625192b");
+        paymentRequestDataDto.setLanguage("FI");
+        paymentRequestDataDto.setPaymentMethod("nordea");
+        paymentRequestDataDto.setPaymentMethodLabel("Nordea");
+
+        OrderWrapper orderWrapper = createDummyOrderWrapper();
+        paymentRequestDataDto.setOrder(orderWrapper);
+
+        String mockPaymentId = PaymentUtil.generatePaymentOrderNumber(orderWrapper.getOrder().getOrderId());
+        Payment mockPayment = mockCreateValidPaymentFromOrder(paymentRequestDataDto, mockPaymentId);
+
+        MvcResult response = this.mockMvc.perform(
+                        post("/payment/paytrail/createFromOrder")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .accept(MediaType.APPLICATION_JSON)
+                                .content(mapper.writeValueAsString(paymentRequestDataDto))
+                )
+                .andDo(print())
+                .andExpect(status().is2xxSuccessful())
+                .andExpect(status().is(201))
+                .andReturn();
+        Payment responseDto = mapper.readValue(response.getResponse().getContentAsString(), Payment.class);
+        Assertions.assertNotNull(responseDto);
+
+        /*
+        * Not able to verify exact paymentId for mock and actual payment.
+        * PaymentId includes orderId and timestamp, so the timestamp of mock paymentId will differ from actual payment.
+        * Here instead we extract orderId from paymentId and verify it matches.
+        */
+        String expectedOrderIdFromPaymentId = mockPayment.getPaymentId().split("_", 2)[0];
+        String responseOrderIdFromPaymentId = responseDto.getPaymentId().split("_", 2)[0];
+
+        Assertions.assertEquals(expectedOrderIdFromPaymentId, responseOrderIdFromPaymentId);
+        Assertions.assertEquals(responseDto.getOrderId(), mockPayment.getOrderId());
+        Assertions.assertEquals(responseDto.getNamespace(), mockPayment.getNamespace());
+        Assertions.assertEquals(responseDto.getUserId(), mockPayment.getUserId());
+        Assertions.assertEquals(responseDto.getPaymentMethod(), mockPayment.getPaymentMethod());
+    }
+
+    @Test
+    public void whenCreateFromOrderWithInvalidPaymentContextThenReturnStatus500() {
+        GetPaymentRequestDataDto paymentRequestDataDto = new GetPaymentRequestDataDto();
+        paymentRequestDataDto.setMerchantId("01fde0e9-82b2-4846-acc0-94291625192b");
```

Comment:
use function mentioned above

Reply:
Here we can't leverage any calls to external services, because this is a unit test and everything is mocked